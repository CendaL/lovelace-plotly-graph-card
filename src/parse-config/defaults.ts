import { parseColorScheme } from "./parse-color-scheme";

export const defaultEntity = {
  entity: "",

  mode: "lines",
  show_value: false,
  line: {
    width: 1,
    shape: "hv",
    color: ({ getFromConfig, entityIdx }) => {
      const color_scheme = parseColorScheme(getFromConfig("color_scheme"));
      return color_scheme[entityIdx % color_scheme.length];
    },
  },
  internal: false,
  offset: "0s",
  // extend_to_present: true unless using statistics. Defined inside parse-config.ts to avoid forward depndency
  unit_of_measurement: ({ meta }) => meta.unit_of_measurement || "",
  name: ({ meta, entityIdx, getFromConfig }) => {
    let name =
      meta.friendly_name || getFromConfig(`entities.${entityIdx}.entity`);
    const attribute = getFromConfig(`entities.${entityIdx}.attribute`);
    if (attribute) name += ` (${attribute}) `;
    return name;
  },
  hovertemplate: ({ getFromConfig, entityIdx }) =>
    `<b>${getFromConfig(
      `entities.${entityIdx}.name`
    )}</b><br><i>%{x}</i><br>%{y} ${getFromConfig(
      `entities.${entityIdx}.unit_of_measurement`
    )}<extra></extra>`,
  yaxis: ({ getFromConfig, entityIdx }) => {
    const units: string[] = [];
    for (let i = 0; i <= entityIdx; i++) {
      const unit = getFromConfig(`entities.${i}.unit_of_measurement`);
      const internal = getFromConfig(`entities.${i}.internal`);
      if (!internal && !units.includes(unit)) units.push(unit);
    }
    const yaxis_idx = units.length;
    return "y" + (yaxis_idx === 1 ? "" : yaxis_idx);
  },
};
export const defaultYaml = {
  title: "",
  hours_to_show: 1,
  refresh_interval: "auto",
  color_scheme: "category10",
  offset: "0s",
  no_theme: false,
  no_default_layout: false,
  disable_pinch_to_zoom: false,
  defaults: {
    entity: {},
    yaxes: {},
  },
  config: {
    displaylogo: false,
    scrollZoom: true,
    modeBarButtonsToRemove: ["resetScale2d", "toImage", "lasso2d", "select2d"],
  },
  layout: {
    xaxis: { autorange: false },
    margin: {
      r: ({ getFromConfig }) => {
        const entities = getFromConfig(`entities`);
        const usesRightAxis = entities.some(({ yaxis }) => yaxis === "y2");
        return usesRightAxis ? 60 : 30;
      },
    },
  },
};
