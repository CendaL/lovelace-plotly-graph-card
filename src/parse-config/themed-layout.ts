export type HATheme = {
  "card-background-color": string;
  "primary-background-color": string;
  "primary-color": string;
  "primary-text-color": string;
  "secondary-text-color": string;
};

const themeAxisStyle = {
  tickcolor: "rgba(127,127,127,.3)",
  gridcolor: "rgba(127,127,127,.3)",
  linecolor: "rgba(127,127,127,.3)",
  zerolinecolor: "rgba(127,127,127,.3)",
};

export default function getThemedLayout(
  haTheme: HATheme
): Partial<Plotly.Layout> {
  return {
    paper_bgcolor: haTheme["card-background-color"],
    plot_bgcolor: haTheme["card-background-color"],
    font: {
      color: haTheme["secondary-text-color"],
      size: 11,
    },
    xaxis: { ...themeAxisStyle },
    yaxis: { ...themeAxisStyle },
    ...Object.fromEntries(
      Array.from({ length: 28 }).map((_, i) => [
        `yaxis${i + 2}`,
        { ...themeAxisStyle },
      ])
    ),
  };
}
