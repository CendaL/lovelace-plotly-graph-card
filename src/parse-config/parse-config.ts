import Cache from "../cache/Cache";
import getThemedLayout, { HATheme } from "./themed-layout";

import merge from "lodash/merge";
import get from "lodash/get";
import { defaultEntity, defaultYaml } from "./defaults";
import { parseTimeDuration } from "../duration/duration";
import { parseStatistics } from "./parse-statistics";
import { HomeAssistant } from "custom-card-helpers";
import filters from "../filters/filters";
import bounds from "binary-search-bounds";
import { has } from "lodash";

// TODO: use Function
const myEval = typeof window != "undefined" ? window.eval : global.eval;

function isObjectOrArray(value) {
  return value !== null && typeof value == "object" && !(value instanceof Date);
}
const w = window as any;

function is$fn(value) {
  return (
    typeof value === "function" ||
    (typeof value === "string" && value.startsWith("$fn"))
  );
}

function removeOutOfRange(data: any, range: [number, number]) {
  const first = bounds.le(data.xs, range[0]);
  if (first > -1) {
    data.xs.splice(0, first);
    data.xs[0] = new Date(range[0]);
    data.ys.splice(0, first);
    data.states.splice(0, first);
    data.statistics.splice(0, first);
  }
  const last = bounds.gt(data.xs, range[1]);
  if (last > -1) {
    data.xs.splice(last);
    data.ys.splice(last);
    data.states.splice(last);
    data.statistics.splice(last);
  }
}

class ConfigParser {
  private partiallyParsedConfig?: any;
  private inputConfig?: any;
  private hass?: HomeAssistant;
  cache = new Cache();
  private busy = false;
  private t_fetch = 0;
  private observed_range: [number, number] = [Date.now(), Date.now()];
  public resetObservedRange() {
    this.observed_range = [Date.now(), Date.now()];
  }
  private getEvaledPath(p: { path: string; callingPath: string }) {
    if (has(this.partiallyParsedConfig, p.path))
      return get(this.partiallyParsedConfig, p.path);

    let value = this.inputConfig;
    for (const key of p.path.split(".")) {
      if (value === undefined) return undefined;
      value = value[key];
      if (is$fn(value)) {
        throw new Error(
          `Since [${p.path}] is a $fn, it has to be defined before [${p.callingPath}]`
        );
      }
    }
    return value;
  }
  private async fetchDataForEntity({
    path,
    fnParam,
  }: {
    parent: object;
    path: string;
    key: string;
    value: any;
    fnParam: any;
  }) {
    path = path.match(/^(entities\.\d+)\./)![1];
    let visible_range = fnParam.getFromConfig("visible_range");
    if (!visible_range) {
      const hours_to_show = fnParam.getFromConfig("hours_to_show");
      const global_offset = parseTimeDuration(fnParam.getFromConfig("offset"));
      const ms = hours_to_show * 60 * 60 * 1000;
      visible_range = [
        +new Date() - ms + global_offset,
        +new Date() + global_offset,
      ] as [number, number];
      this.partiallyParsedConfig.visible_range = visible_range;
    }
    this.observed_range[0] = Math.min(this.observed_range[0], visible_range[0]);
    this.observed_range[1] = Math.max(this.observed_range[1], visible_range[1]);
    const statisticsParams = parseStatistics(
      visible_range,
      fnParam.getFromConfig(path + ".statistic"),
      fnParam.getFromConfig(path + ".period")
    );
    const attribute = fnParam.getFromConfig(path + ".attribute") as
      | string
      | undefined;
    const fetchConfig = {
      entity: fnParam.getFromConfig(path + ".entity"),
      ...(statisticsParams ? statisticsParams : attribute ? { attribute } : {}),
    };
    const offset = parseTimeDuration(fnParam.getFromConfig(path + ".offset"));

    const range_to_fetch = [
      visible_range[0] - offset,
      visible_range[1] - offset,
    ];
    const t0 = performance.now();
    const fetch_mask = fnParam.getFromConfig("fetch_mask");
    const data =
      fetch_mask[fnParam.entityIdx] === false // also fetch if it is undefined. This means the entity is new
        ? this.cache.getData(fetchConfig)
        : await this.cache.fetch(range_to_fetch, fetchConfig, this.hass!);
    const extend_to_present =
      fnParam.getFromConfig(path + ".extend_to_present") ?? !statisticsParams;

    data.xs = data.xs.map((x) => new Date(+x + offset));

    removeOutOfRange(data, this.observed_range);
    if (extend_to_present && data.xs.length > 0) {
      const last_i = data.xs.length - 1;
      const now = Math.min(this.observed_range[1], Date.now());
      data.xs.push(new Date(Math.min(this.observed_range[1], now + offset)));
      data.ys.push(data.ys[last_i]);
      if (data.states.length) data.states.push(data.states[last_i]);
      if (data.statistics.length) data.statistics.push(data.statistics[last_i]);
    }
    this.t_fetch += performance.now() - t0;
    fnParam.data = {
      ...data,
      meta: this.hass?.states[fetchConfig.entity]?.attributes || {},
      vars: fnParam.vars,
      hass: this.hass!,
    };
  }
  private async fetchDataForEntityIfNecessary(p: {
    parent: object;
    path: string;
    key: string;
    value: any;
    fnParam: any;
  }) {
    const isInsideEntity = !!p.path.match(/^entities\.\d+\./);
    const isInsideFetchParamNode = !!p.path.match(
      /^entities\.\d+\.(entity|attribute|offset|statistic|period)/
    );
    const alreadyFetchedData = p.fnParam.data;
    const isFilters = p.path.match(/^entities\.\d+\.filters\.\d+$/);
    if (
      isInsideEntity &&
      !isInsideFetchParamNode &&
      !alreadyFetchedData &&
      (is$fn(p.value) || isFilters)
    )
      await this.fetchDataForEntity(p);
  }
  private async evalNode({
    parent,
    path,
    key,
    value,
    fnParam,
  }: {
    parent: object;
    path: string;
    key: string;
    value: any;
    fnParam: any;
  }) {
    w.i++;
    console.log(path);
    if (path.match(/^defaults$/)) return;
    if (path.match(/^entities\.\d+$/)) {
      fnParam.entityIdx = key;
    }
    fnParam.getFromConfig = (aPath: string) =>
      this.getEvaledPath({ path: aPath, callingPath: path });
    await this.fetchDataForEntityIfNecessary({
      parent,
      path,
      key,
      value,
      fnParam,
    });

    if (typeof value === "string" && value.startsWith("$fn")) {
      value = myEval(value.slice(3));
    }

    if (typeof value === "function") {
      /**
       * Allowing functions that return functions makes it very slow
       * This is mostly because of customdata, which returns an array as large as the cached data,
       * and the fact that awaits are expensive.
       */

      parent[key] = value = value(fnParam);
    } else if (isObjectOrArray(value)) {
      const me = Array.isArray(value) ? [] : {};
      parent[key] = me;
      for (const [childKey, childValue] of Object.entries(value)) {
        await this.evalNode({
          parent: me,
          path: `${path}.${childKey}`,
          key: childKey,
          value: childValue,
          fnParam,
        });
      }
    } else {
      parent[key] = value;
    }

    // we're now on the way back of traversal, `value` is fully evaluated

    if (path.match(/^entities\.\d+$/)) {
      if (!fnParam.data) {
        await this.fetchDataForEntity({
          parent,
          path,
          key,
          value,
          fnParam,
        });
      }
      const me = parent[key];
      if (!me.x) me.x = fnParam.data.xs;
      if (!me.y) me.y = fnParam.data.ys;
      if (me.x.length === 0 && me.y.length === 0) {
        /*
        Traces with no data are removed from the legend by plotly. 
        Setting them to have null element prevents that.
        */
        me.x = [new Date()];
        me.y = [null];
      }
      delete fnParam.data;
      delete fnParam.entityIdx;
    }

    if (path.match(/^entities\.\d+\.filters\.\d+$/)) {
      evalFilter({ parent, path, key, value, fnParam });
    }
    if (path.match(/^entities$/)) {
      parent[key] = parent[key].filter(({ internal }) => !internal);
      const entities = parent[key];
      const count = entities.length;
      // Preserving the original sequence of real_traces is important for `fill: tonexty`
      // https://github.com/dbuezas/lovelace-plotly-graph-card/issues/87
      for (let i = 0; i < count; i++) {
        const trace = entities[i];
        if (trace.show_value) {
          trace.legendgroup ??= "group" + i;
          entities.push({
            texttemplate: `%{y:.2~f}%{customdata.unit_of_measurement}`, // here so it can be overwritten
            ...trace,
            cliponaxis: false, // allows the marker + text to be rendered above the right y axis. See https://github.com/dbuezas/lovelace-plotly-graph-card/issues/171
            mode: "text+markers",
            showlegend: false,
            hoverinfo: "skip",
            textposition: "middle right",
            marker: {
              color: trace.line?.color,
            },
            textfont: {
              color: trace.line?.color,
            },
            x: trace.x.slice(-1),
            y: trace.y.slice(-1),
          });
        }
      }
    }
  }
  async update(input: {
    raw_config: any;
    hass: HomeAssistant;
    cssVars: HATheme;
  }) {
    w.i = 0;
    const t0 = performance.now();
    this.t_fetch = 0;
    /*
      TODOs:
      * REMEMBER TO PASS visible entities
      * const visibleEntities = this.parsed_config.entities.filter(
        (_, i) => this.contentEl.data[i]?.visible !== "legendonly"
      );
      * remember to pass observed_range (all viewed ranges) to cap the range of the data
    */
    if (this.busy) throw new Error("ParseConfig was updated while busy");
    this.busy = true;
    this.hass = input.hass;
    const old_uirevision = this.partiallyParsedConfig?.layout?.uirevision;
    let config = JSON.parse(JSON.stringify(input.raw_config));

    // 1st pass: add defaults
    config = merge({}, config, defaultYaml, config);
    for (let i = 1; i < 31; i++) {
      const yaxis = "yaxis" + (i == 1 ? "" : i);
      config.layout[yaxis] = merge(
        {},
        config.layout[yaxis],
        config.defaults?.yaxes,
        config.layout[yaxis]
      );
    }
    config.entities = config.entities.map((entity) => {
      if (typeof entity === "string") entity = { entity };
      entity.entity ??= "";
      const [oldAPI_entity, oldAPI_attribute] = entity.entity.split("::");
      if (oldAPI_attribute) {
        entity.entity = oldAPI_entity;
        entity.attribute = oldAPI_attribute;
      }
      entity = merge(
        {},
        entity,
        defaultEntity,
        config.defaults?.entity,
        entity
      );
      return entity;
    });

    // 2nd pass: evaluate functions
    const fnParam = {
      vars: {},
      hass: input.hass,
    };
    this.partiallyParsedConfig = {};
    this.inputConfig = config;
    for (const [key, value] of Object.entries(config)) {
      await this.evalNode({
        parent: this.partiallyParsedConfig,
        path: key,
        key: key,
        value,
        fnParam,
      });
    }

    // 3rd pass: decorate
    // TODO: do this in defaults
    const isBrowsing = !!input.raw_config.visible_range;
    const yAxisTitles = Object.fromEntries(
      this.partiallyParsedConfig.entities.map(
        ({ unit_of_measurement, yaxis }) => [
          "yaxis" + yaxis.slice(1),
          { title: unit_of_measurement },
        ]
      )
    );
    merge(
      this.partiallyParsedConfig.layout,
      getThemedLayout(
        input.cssVars,
        this.partiallyParsedConfig.no_theme,
        this.partiallyParsedConfig.no_default_layout
      ),
      {
        xaxis: {
          range: this.partiallyParsedConfig.visible_range,
        },
        //changing the uirevision triggers a reset to the axes
        uirevision: isBrowsing ? old_uirevision : Math.random(),
      },
      this.partiallyParsedConfig.no_default_layout ? {} : yAxisTitles,
      this.partiallyParsedConfig.layout
    );
    this.busy = false;

    const t_total = performance.now() - t0;
    console.table({
      t_total,
      t_fetch: this.t_fetch,
      t_parse: t_total - this.t_fetch,
    });
    console.log(w.i);
    return this.partiallyParsedConfig;
  }
}

export { ConfigParser };

function evalFilter(input: {
  parent: object;
  path: string;
  key: string;
  value: any;
  fnParam: any;
}) {
  const obj = input.value;
  let filterName: string;
  let config: any = null;
  if (typeof obj === "string") {
    filterName = obj;
  } else {
    filterName = Object.keys(obj)[0];
    config = Object.values(obj)[0];
  }
  const filter = filters[filterName];
  if (!filter) {
    throw new Error(`Filter '${filterName} must be [${Object.keys(filters)}]`);
  }
  const filterfn = config === null ? filter() : filter(config);
  try {
    const fnParam = input.fnParam;
    fnParam.data.vars = fnParam.vars;
    input.fnParam.data = {
      ...input.fnParam.data,
      ...filterfn(input.fnParam.data),
    };
    fnParam.vars = fnParam.data.vars;
  } catch (e) {
    console.error(e);
    throw new Error(`Error in filter: ${e}`);
  }
}
