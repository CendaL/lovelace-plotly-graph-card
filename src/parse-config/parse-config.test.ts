import { ConfigParser } from "./parse-config";

const the_config = {
  refresh_interval: '$fn ({vars, hass}) => "auto"',
  offset: '$fn ({vars, hass}) => {vars.a_var = "from offset"; return "5s";}',
  hours_to_show: "$fn ({vars, hass, xs, ys}) => 5",
  color_scheme: `$fn ()=>0`,

  entities: [
    {
      entity: "sensor.temperature",
      attribute: '$fn ({vars, hass}) => "temperature"',
      offset: "5s",
      unit_of_measurement: "W",
      text: "$fn ({vars, hass, xs, ys}) => { vars.ys = ys; return ys.map(y => 'y=${y}')}",
      filters: [
        {
          add: "$fn ({vars, hass}) => 5",
        },
      ],
    },
  ],
  defaults: {
    entity: {
      internal: "$fn ({vars, hass, xs, ys}) => ys.map(y => y%2)",
      show_value: true,
    },
    yaxes: {
      autorange: "$fn ({vars, hass}) => true",
      fixedrange: false,
    },
  },
  title: "$fn ({vars, hass}) => vars.a_var",
  layout: {
    legend: { traceorder: "normal" },
  },
  disable_pinch_to_zoom: "$fn ({vars, hass}) => vars.a_var",
  stuff_coming_from_fetched_data: "$fn ({vars, hass}) => vars",
};

const parser = new ConfigParser();
const hass = {};
const cssVars = {
  "--card-background-color": "--card-background-color",
  "--primary-background-color": "--primary-background-color",
  "--primary-color": "--primary-color",
  "--primary-text-color": "--primary-text-color",
  "--secondary-text-color": "--secondary-text-color",
};

describe("ConfigParser", () => {
  it("Can parse", async () => {
    await parser.update({
      raw_config: the_config,
      hass,
      cssVars,
    });
    expect(parser.config).toMatchSnapshot();
  });
});
console.log(JSON.stringify(parser.config, null, 2));
