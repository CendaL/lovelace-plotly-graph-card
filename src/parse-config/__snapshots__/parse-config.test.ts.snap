// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigParser Can parse 1`] = `
{
  "color_scheme": 0,
  "config": {
    "displaylogo": false,
    "modeBarButtonsToRemove": [
      "resetScale2d",
      "toImage",
      "lasso2d",
      "select2d",
    ],
    "scrollZoom": true,
  },
  "disable_pinch_to_zoom": "from offset",
  "entities": [
    {
      "attribute": "temperature",
      "entity": "sensor.temperature",
      "filters": [
        {
          "add": 5,
        },
      ],
      "hovertemplate": "<b>%{customdata.name}</b><br><i>%{x}</i><br>%{y}%{customdata.unit_of_measurement}<extra></extra>",
      "internal": [
        1,
        0,
        1,
        0,
      ],
      "line": {
        "color": "#7fc97f",
        "shape": "hv",
        "width": 1,
      },
      "mode": "lines",
      "offset": "5s",
      "show_value": true,
      "text": [
        "y=\${y}",
        "y=\${y}",
        "y=\${y}",
        "y=\${y}",
      ],
      "unit_of_measurement": "W",
      "x": [
        1,
        2,
        3,
        4,
      ],
      "y": [
        21,
        22,
        23,
        24,
      ],
    },
  ],
  "hours_to_show": 5,
  "layout": {
    "dragmode": "pan",
    "font": {
      "color": "--secondary-text-color",
      "size": 11,
    },
    "height": 285,
    "legend": {
      "bgcolor": "transparent",
      "orientation": "h",
      "traceorder": "normal",
      "x": 0,
      "y": 1,
      "yanchor": "bottom",
    },
    "margin": {
      "b": 50,
      "l": 60,
      "r": 60,
      "t": 0,
    },
    "modebar": {
      "orientation": "v",
    },
    "paper_bgcolor": "--card-background-color",
    "plot_bgcolor": "--card-background-color",
    "title": {
      "pad": {
        "t": 15,
      },
      "y": 1,
    },
    "xaxis": {
      "autorange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "linecolor": "rgba(127,127,127,.3)",
      "tickcolor": "rgba(127,127,127,.3)",
      "type": "date",
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "linecolor": "rgba(127,127,127,.3)",
      "tickcolor": "rgba(127,127,127,.3)",
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis10": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis11": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis12": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis13": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis14": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis15": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis16": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis17": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis18": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis19": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis2": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": true,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis20": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis21": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis22": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis23": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis24": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis25": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis26": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis27": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis28": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis29": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis3": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis30": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis4": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis5": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis6": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis7": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis8": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
    "yaxis9": {
      "autorange": true,
      "fixedrange": false,
      "gridcolor": "rgba(127,127,127,.3)",
      "layer": "below traces",
      "linecolor": "rgba(127,127,127,.3)",
      "overlaying": "y",
      "showgrid": false,
      "side": "right",
      "tickcolor": "rgba(127,127,127,.3)",
      "visible": false,
      "zerolinecolor": "rgba(127,127,127,.3)",
    },
  },
  "minimal_response": true,
  "no_default_layout": false,
  "no_theme": false,
  "offset": "5s",
  "refresh_interval": "auto",
  "significant_changes_only": false,
  "stuff_coming_from_fetched_data": {
    "a_var": "from offset",
    "ys": [
      21,
      22,
      23,
      24,
    ],
  },
  "title": "from offset",
}
`;
